src/ln/msgs.rs

```diff
@@ -1200,112 +1265,76 @@ impl<R: Read> Readable<R> for UnsignedNodeAnnouncement {
+		let mut addresses: Vec<NetAddress> = Vec::with_capacity(addr_len as usize);
+		let mut known_addr_len = 0;
+		let mut excess = false;
+		let mut excess_byte = 0;
+		for _ in 0..addr_len {
+			match Readable::read(r) {
+				Ok(addr) => {
+					match addr {
+						NetAddress::IPv4 { .. } => {
+							if addresses.len() > 0 {
+								return Err(DecodeError::ExtraAddressesPerType);
+							}
 						},
-						port: Readable::read(r)?,
-					});
-					addr_readpos += 1 + 6
-				},
-				2 => {
-					if addresses.len() > 1 || (addresses.len() == 1 && addresses[0].get_id() != 1) {
-						return Err(DecodeError::ExtraAddressesPerType);
-					}
-					if addrlen < addr_readpos + 1 + 18 {
-						return Err(DecodeError::BadLengthDescriptor);
-					}
-					addresses.push(NetAddress::IPv6 {
-						addr: {
-							let mut addr = [0; 16];
-							r.read_exact(&mut addr)?;
-							addr
+						NetAddress::IPv6 { .. } => {
+							if addresses.len() > 1 || (addresses.len() == 1 && addresses[0].get_id() != 1) {
+								return Err(DecodeError::ExtraAddressesPerType);
+							}
 						},
-						port: Readable::read(r)?,
-					});
-					addr_readpos += 1 + 18
-				},
-				3 => {
-					if addresses.len() > 2 || (addresses.len() > 0 && addresses.last().unwrap().get_id() > 2) {
-						return Err(DecodeError::ExtraAddressesPerType);
-					}
-					if addrlen < addr_readpos + 1 + 12 {
-						return Err(DecodeError::BadLengthDescriptor);
-					}
-					addresses.push(NetAddress::OnionV2 {
-						addr: {
-							let mut addr = [0; 10];
-							r.read_exact(&mut addr)?;
-							addr
+						NetAddress::OnionV2 { .. } => {
+							if addresses.len() > 2 || (addresses.len() > 0 && addresses.last().unwrap().get_id() > 2) {
+								return Err(DecodeError::ExtraAddressesPerType);
+							}
 						},
-						port: Readable::read(r)?,
-					});
-					addr_readpos += 1 + 12
-				},
-				4 => {
-					if addresses.len() > 3 || (addresses.len() > 0 && addresses.last().unwrap().get_id() > 3) {
-						return Err(DecodeError::ExtraAddressesPerType);
+						NetAddress::OnionV3 { .. } => {
+							if addresses.len() > 3 || (addresses.len() > 0 && addresses.last().unwrap().get_id() > 3) {
+								return Err(DecodeError::ExtraAddressesPerType);
+							}
+						}
 					}
-					if addrlen < addr_readpos + 1 + 37 {
+					if addr_len < known_addr_len + 1 + NetAddress::len(&addr) {
 						return Err(DecodeError::BadLengthDescriptor);
 					}
-					addresses.push(NetAddress::OnionV3 {
-						ed25519_pubkey: Readable::read(r)?,
-						checksum: Readable::read(r)?,
-						version: Readable::read(r)?,
-						port: Readable::read(r)?,
-					});
-					addr_readpos += 1 + 37
+					known_addr_len += (1 + NetAddress::len(&addr)) as u16;
+					addresses.push(addr);
 				},
-				_ => { excess = 1; break; }
+				Err(DecodeError::UnknownVersion { byte }) => {
+					excess = true;
+					excess_byte = byte;
+					break;
+				},
+				Err(DecodeError::ShortRead) => return Err(DecodeError::BadLengthDescriptor),
```

Comment:
There, as serialization is implemented for NetAddress now, a ShortRead bring back by deserialization field failure should be translated as a BadLengthDescriptor as address descriptor isn't accurate seems to me (or we can change msg_node_announcement_target constraints)